erpnext/stock/doctype/delivery_note/delivery_note.py

```diff
@@ -628,3 +634,24 @@ def make_sales_return(source_name, target_doc=None):
 def update_delivery_note_status(docname, status):
 	dn = frappe.get_doc("Delivery Note", docname)
 	dn.update_status(status)
+
+@frappe.whitelist()
+def email_coas(docname):
+	delivery_note = frappe.get_doc("Delivery Note", docname).as_dict()
+	fid = {}
+	attachments = []
+	if not delivery_note.get("contact_email"):
+		frappe.throw(_("No contact email found"))
+
+	for item in delivery_note.get("items"):
+		if not item.certificate_of_analysis:
+			frappe.msgprint(_("No Certificate of Analysis attached"))
```

Comment:
You should only do this once per document - check all items at once to see if there's any attached CoAs

```suggestion
		coas = [item.get("certificate_of_analysis") for item in delivery_note.get("items")
			if item.get("certificate_of_analysis")]
		if not coas:
			frappe.msgprint(_("No Certificates of Analysis attached"))
			return
```